### PR TITLE
Added flip function (to a tensor up to 5D along an arbitrary dimension).

### PR DESCRIPTION
**Pull request by jonathantompson**
The c function doesn't support in-place flips, but I check for this condition in the C code.  I'm using this for a scientific application, however you could also use it to swap RGB -> BGR which might be handy for some folks:

```lua
im = image.lena()
image.display(im)
image.display(image.flip(im, 1))  -- Swap R with B chan
```

### CODE REVIEW

**Comment by soumith** [comment]
will fail for non-contiguous inputs right :)

**Comment by soumith** [comment]
oh actualy looks like you used the strides, so nevermind it'll work

**Comment by jonathantompson** [comment]
Yes, but so does vflip and hflip so I figured I would leave it.

On Tue, Apr 14, 2015 at 11:58 AM, Soumith Chintala <notifications@github.com

**Comment by jonathantompson** [comment]
Never mind, hflip does work on non-contiguous inputs :-)  That's funny...
I had convinced myself at some point it didn't.  Ohh well.

**Comment by soumith** [comment]
can you add a test-case for non-contiguous, your code looks like it should work for it.

**Comment by jonathantompson** [comment]
OK, I added a test but it fails.  FYI: hflip and vflip also fail.  Do you want me to add asserts in hflip, vflip and flip checking that the input is contiguous?

**Comment by soumith** [comment]
Yes those asserts would be fantastic!!!

**Comment by jonathantompson** [comment]
Yeah, here's a quick sanity check for hflip:

```
require 'image'
input = torch.rand(6,6)
input = input:transpose(1,2)
assert(not input:isContiguous())
out1 = image.hflip(input)
out2 = image.hflip(input:contiguous())
err =  out2 - out1
print(err)
```

**Comment by jonathantompson** [comment]
OK, done.  Since this changes existing code, can you please do a quick code review of the changes in init.lua (there's only a few lines in hflip and vlip)?

Playing more with hflip, it seems that "some" non-contiguous tensors work and some do not.  Non-continuity from `:select()` works, but `:transpose()` causes it to fail.  I think failing on all non-contiguous tensors is the correct behaviour until the transpose case is fixed...  However, I'd rather do that in a separate PR.

File: init.lua [hunk 1]
```diff
@@ -856,6 +876,69 @@ end
...
 ----------------------------------------------------------------------
+-- flip (specify dimension, up to 5D tensor)
+--
+local function flip(...)
+   local dst,src
+   local args = {...}
+   if select('#',...) == 3 then
+      dst = args[1]
+      src = args[2]
+      flip_dim = args[3]
```
Comment (soumith) [inline init.lua]: you are leaking a global variable here. flip_dim

Comment (jonathantompson) [inline init.lua]: Wow, they don't call you "dead-eye Chintala" for nothing :-)  Good catch.

I fixed it, rebased and pushed.

On Tue, Apr 14, 2015 at 11:49 PM, Soumith Chintala <notifications@github.com

**Comment by soumith** [comment]
looks good to me! thanks jono

**Comment by jonathantompson** [comment]
Quick question.  For the old hflip / vflip on my machine the following snippet returned non-zero error.  Does it work for you (ie is err zero)?

```lua
require 'image'
input = torch.rand(6,6)
input = input:transpose(1,2)
assert(not input:isContiguous())
out1 = image.hflip(input)
out2 = image.hflip(input:contiguous())
err =  out2 - out1
print(err)
```

**Comment by jonathantompson** [comment]
What do you mean "wrong!"?  Am I mis-understanding something?

The above snippit gives me:

```
t7> =out1
 0.3173  0.9720  0.7057  0.0146  0.0713  0.6670
 0.3956  0.5464  0.6365  0.4492  0.3873  0.1803
 0.3049  0.7097  0.2206  0.9484  0.8096  0.5397
 0.3628  0.5579  0.6938  0.9875  0.5544  0.0341
 0.7052  0.4693  0.8795  0.3828  0.9875  0.3214
 0.3606  0.1706  0.1223  0.7818  0.0205  0.0355
[torch.DoubleTensor of dimension 6x6]
```

and

```
t7> =out2
 0.0355  0.3214  0.0341  0.5397  0.1803  0.6670
 0.0205  0.9875  0.5544  0.8096  0.3873  0.0713
 0.7818  0.3828  0.9875  0.9484  0.4492  0.0146
 0.1223  0.8795  0.6938  0.2206  0.6365  0.7057
 0.1706  0.4693  0.5579  0.7097  0.5464  0.9720
 0.3606  0.7052  0.3628  0.3049  0.3956  0.3173
[torch.DoubleTensor of dimension 6x6]
```

Why would these be different?  Or am I missing something?

**Comment by Aysegul** [comment]
Hi Jonathan, you are right. The problem is when you have 2D input, hlip function resizes the input
from init.lua
```lua
 if src:nDimension() == 2 then
      src = src:new():resize(1,src:size(1),src:size(2))
   end
```

it changes the stride and reverts the transpose

```lua
 input = torch.rand(6,6)
 input = input:transpose(1,2)
 print(input)
 
 input = input:new():resize(1, 6, 6)
print(input)
```
they are different. Not sure, what is the best way to solve that.

**Comment by jonathantompson** [comment]
I see.  Yeah, without digging into it earlier to figure out the cause I decided I would add the contiguity check in the last PR as a safe guard until it gets fixed.  Sorry for folks whos 3D tensor flips are now failing.  Perhaps someone can submit another PR with a modified assert.

**Pull request merged**